src/core.php

```diff
@@ -1101,3 +1226,29 @@ public function __construct(:x:composable-element $that, int $index) {
 interface XHPUnsafeRenderable extends XHPChild {
   public function toHTMLString();
 }
+
+/**
+ * INCREDIBLY AWESOME: Specify an element as awaitable on render.
+ *
+ * This allows you to use generators inside your XHP objects. For
+ * instance, you could fetch data inside your XHP elements using await and
+ * the calls to the DB would be batched together when the element is rendered.
+ */
+interface XHPAwaitable {
+  // protected function genRender(): Awaitable<XHPRoot>
+}
+
+trait XHPGenerator {
+
+  require extends :x:element;
+  require implements XHPAwaitable;
+
+  abstract protected function genRender(): Awaitable<XHPRoot>;
```

Comment:
@sgolemon: was there a decision on what genFoo methods should be called externally?

@Swahvay: more generally, this shouldn't be called 'Generator', as they're now an entirely separate concept from awaitables